resolve.c

```diff
@@ -71,6 +63,37 @@ static char *resolveHostname(const char *addr)
 	return hostname;
 }
 
+// Resolve upstream destination host names
+static size_t resolveAndAddHostname(size_t ippos, size_t oldnamepos)
+{
+	// Get IP and host name strings
+	const char* ipaddr = getstr(ippos);
+	const char* oldname = getstr(oldnamepos);
+
+	// Important: Don't hold a lock while resolving as the main thread
+	// (dnsmasq) needs to be operable during the call to resolveHostname()
```

Comment:
This function needs to lock shared memory just as it did before it was a function. Memory is read by `getstr` and written to by `addstr`, but should not be locked when calling `resolveHostname`.

Reply:
The string buffer is only every appended so `getstr()` can be safely called without memory actions taking place during its run. You are right, however, as the string buffer might get extended just when we run `getstr()` which is something we obviously want to avoid.